server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -352,24 +355,73 @@ public void merge(IndexMetadata indexMetadata, MergeReason reason) {
-        if (currentMapper != null && currentMapper.mappingSource().equals(mappingSource)) {
-            return currentMapper;
+        if (currentMapper != null) {
+            if (mappingSources.size() == 1 && currentMapper.mappingSource().equals(mappingSources.get(0))) {
+                return currentMapper;
+            }
+            if (currentMapper.mapping().getRoot().subobjectsExplicit()) {
+                subobjects = Explicit.explicitBoolean(currentMapper.mapping().getRoot().isEnabled());
+            }
+        }
+
+        List<Mapping> parsedMappings = new LinkedList<>();
+        for (CompressedXContent mappingSource : mappingSources) {
+            Mapping mapping = parseMapping(type, mappingSource);
+            if (mapping.getRoot().subobjectsExplicit()) {
+                boolean explicitSubobjects = mapping.getRoot().subobjects();
+                if (subobjects != null && subobjects.value() != explicitSubobjects) {
+                    throw new MapperParsingException("Failed to parse mappings: contradicting subobjects settings provided");
+                }
+                subobjects = Explicit.explicitBoolean(explicitSubobjects);
+            }
+            parsedMappings.add(mapping);
+        }
+
+        if (subobjects != null && subobjects.value() != ObjectMapper.Defaults.SUBOBJECTS.value()) {
+            // we need to reparse as parsing is affected by the subobjects setting
+            parsedMappings.clear();
+            for (CompressedXContent mappingSource : mappingSources) {
+                parsedMappings.add(parseMapping(type, mappingSource, subobjects));
+            }
+        }
+
+        DocumentMapper ret = null;
+        for (Mapping parsedMapping : parsedMappings) {
+            ret = doMerge(reason, parsedMapping);
         }
-        synchronized (this) {
-            Mapping incomingMapping = parseMapping(type, mappingSource);
-            Mapping mapping = mergeMappings(this.mapper, incomingMapping, reason);
-            // TODO: In many cases the source here is equal to mappingSource so we need not serialize again.
-            // We should identify these cases reliably and save expensive serialization here
-            DocumentMapper newMapper = newDocumentMapper(mapping, reason, mapping.toCompressedXContent());
-            if (reason == MergeReason.MAPPING_UPDATE_PREFLIGHT) {
-                return newMapper;
+        return ret;
+    }
+
+    public DocumentMapper merge(String type, CompressedXContent mappingSource, MergeReason reason) {
+        Explicit<Boolean> subobjects = null;
+        final DocumentMapper currentMapper = this.mapper;
+        if (currentMapper != null) {
+            if (currentMapper.mappingSource().equals(mappingSource)) {
+                return currentMapper;
+            }
+            if (currentMapper.mapping().getRoot().subobjectsExplicit()) {
+                subobjects = Explicit.explicitBoolean(currentMapper.mapping().getRoot().isEnabled());
```

Comment:
We probably don't need this as the current mapper's root already has the correct `subobjects` setting.